src/rgw/rgw_rest_sts.cc

```diff
@@ -223,137 +224,264 @@ WebTokenEngine::get_cert_url(const string& iss, const DoutPrefixProvider *dpp, o
+    throw;
+  }
+  catch (...) {
+    ldpp_dout(dpp, 0) << "Signature validation using x5c failed" << dendl;
+    throw;
+  }
+  ldpp_dout(dpp, 10) << "Verified signature using x5c certificate "<< dendl;
+}
+
+void
+WebTokenEngine::validate_signature_using_n_e(const DoutPrefixProvider* dpp, const jwt::decoded_jwt& decoded, const string& n, const string& e, const string &algorithm) const
+{
+  try {
+    if (algorithm == "RS256") {
+      auto verifier = jwt::verify()
+                  .allow_algorithm(jwt::algorithm::rs256().setModulusAndExponent(n,e));
+      verifier.verify(decoded);
+    } else if (algorithm == "RS384") {
+      auto verifier = jwt::verify()
+                  .allow_algorithm(jwt::algorithm::rs384().setModulusAndExponent(n,e));
+      verifier.verify(decoded);
+    } else if (algorithm == "RS512") {
+      auto verifier = jwt::verify()
+                  .allow_algorithm(jwt::algorithm::rs512().setModulusAndExponent(n,e));
+      verifier.verify(decoded);
+    }
+  } catch (std::runtime_error& e) {
+    ldpp_dout(dpp, 0) << "Signature validation using n, e failed: " << e.what() << dendl;
+    throw;
+  }
+  catch (...) {
+    ldpp_dout(dpp, 0) << "Signature validation n, e failed" << dendl;
+  }
+  ldpp_dout(dpp, 10) << "Verified signature using n and e"<< dendl;
+  return;
 }
 
 void
 WebTokenEngine::validate_signature(const DoutPrefixProvider* dpp, const jwt::decoded_jwt& decoded, const string& algorithm, const string& iss, const vector<string>& thumbprints, optional_yield y) const
 {
   if (algorithm != "HS256" && algorithm != "HS384" && algorithm != "HS512") {
-    string cert_url = get_cert_url(iss, dpp, y);
-    if (cert_url.empty()) {
-      throw -EINVAL;
+    //Get x5c, x5u, kid, x5t from header
+    //if x5c present use it
+    //if x5u is present, get cert
+    //Use kid or x5t to locate a valid cert from JWK
+    string kid, x5t, x5t256;
+    if (decoded.has_header_claim("x5c")) {
+      auto &claim = decoded.get_header_claim("x5c").as_array();
+      vector<string> x5c;
+      for (auto& it : claim) {
+        x5c.emplace_back(it.to_str());
+      }
+      validate_signature_using_cert(dpp, decoded, algorithm, x5c, thumbprints);
+      return;
+    }
+    if (decoded.has_header_claim("x5u")) {
+      string x5u_url = decoded.get_header_claim("x5u").as_string();
+      vector<string> x5c = get_x5c_certs_from_x5u_url(dpp, x5u_url, y);
+      validate_signature_using_cert(dpp, decoded, algorithm, x5c, thumbprints, false);
+      return;
+    }
+    if (decoded.has_key_id()) {
+      kid = decoded.get_key_id();
+    }
+    if (decoded.has_header_claim("x5t")) {
+      x5t = decoded.get_header_claim("x5t").as_string();
+    }
+    if (decoded.has_header_claim("x5t#256")) {
+      x5t256 = decoded.get_header_claim("x5t#256").as_string();
     }
 
+    string jwks_url = get_jwks_url(iss, dpp, y);
+    if (jwks_url.empty()) {
+      throw -EINVAL;
```

Comment:
At some point I want to find all the places that are expecting to catch integers and make them stop. (Though this isn't really your problem since it was like that when you got it.)